lib/internal/worker/locks.js

```diff
@@ -0,0 +1,526 @@
+        this.unlockMutex();
+      },
+      release: () => {
+        this.lockMutex();
+        for (let i = 0; i < this.state.held.length; i++) {
+          if (this.state.held[i].id === request.id) {
+            this.state.held.splice(i, 1);
+            break;
+          }
+        }
+        this.resolveMap.delete(request.id);
+        this.processLockRequestQueue();
+        this.unlockMutex();
+      }
+    };
+  }
+
+  processLockRequestQueue() {
+    let didModify = false;
+    let hasOwnPendingRequests = false;
+    // For each request of queue:
+    for (let i = 0; i < this.state.pending.length;) {
+      const request = this.state.pending[i];
+      if (request.clientId !== this.clientId) {
+        i++;
+        continue;
+      }
+      if (!this.isGrantable(request)) {
+        hasOwnPendingRequests = true;
+        i++;
+        continue;
+      }
+      // If request is grantable, then run these steps:
+      // Append lock to origin's held lock set.
+      this.state.pending.splice(i, 1);
+      this.state.held.push(request);
+      this.availableMap.get(request.id).resolve();
+      this.availableMap.delete(request.id);
+      didModify = true;
+    }
+    for (let i = 0; i < this.state.stolen.length;) {
+      const request = this.state.stolen[i];
+      if (request.clientId !== this.clientId) {
+        i++;
+        continue;
+      }
+
+      // Reject lock's released promise with an "AbortError" DOMException.
+      this.resolveMap.get(request.id).reject(
+        new DOMException('The operation was aborted', 'AbortError'));
+      this.resolveMap.delete(request.id);
+      this.state.stolen.splice(i, 1);
+      didModify = true;
+    }
+    // Let the event loop spin in case there are any pending requests pertaining
+    // to this thread.
+    if (hasOwnPendingRequests) {
+      this.channel.ref();
+    } else {
+      this.channel.unref();
+    }
+    return didModify;
+  }
+
+  isGrantable(request) {
+    // A lock request request is said to be grantable if the following steps
+    // return true:
+    // If mode is "exclusive", then return true if all of the following
+    // conditions are true, and false otherwise:
+    // - No lock in held has a name that equals name
+    // - No lock request in queue earlier than request exists.
+    // Otherwise, mode is "shared"; return true if all of the following
+    // conditions are true, and false otherwise:
+    // - No lock in held has mode "exclusive" and has a name that equals name.
+    // - No lock request in queue earlier than request exists.
+    for (const lock of this.state.held) {
+      if (lock.name === request.name && (!lock.shared || !request.shared)) {
+        return false;
+      }
+    }
+    for (const lock of this.state.pending) {
+      if (lock.id < request.id &&
+          lock.name === request.name && (!lock.shared || !request.shared)) {
+        return false;
+      }
+    }
+    return true;
+  }
+
+  onmessage(message) {
```

Comment:
Any reason this looks like an event handler (`onmessage`)?

Reply:
I mean, yeah, it’s the handler for `'message'` events on the underlying channel.